### PR TITLE
Simplify ssh known hosts handling

### PR DESCRIPTION
**Pull request by Gundersanne**
As this simplifies the known hosts logic a little it relies on libssh checking both local and global known_hosts files in `ssh_session_has_known_hosts_entry`:

- [x] https://bugs.libssh.org/T107 
- [x] PR #10318

### CODE REVIEW

**Comment by Gundersanne** [comment]
Libssh just did a release, so we should be ready to go after the next refreshes.

**Comment by martinpitt** [comment]
Changing the blocks to #10318 instead, this will already require image refreshes.

File: src/ssh/cockpitsshrelay.c [hunk 1]
```diff
@@ -560,8 +560,8 @@ session_has_known_host_in_file (const gchar *file,
 #endif
   /* HACK: https://bugs.libssh.org/T108 */
   if (!file)
-    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == SSH_OK);
-  g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_GLOBAL_KNOWNHOSTS, file) == SSH_OK);
+    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == 0);
+  g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_GLOBAL_KNOWNHOSTS, file) == 0);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Of course, will do

File: src/ssh/cockpitsshrelay.c [hunk 2]
```diff
@@ -556,12 +556,12 @@ session_has_known_host_in_file (const gchar *file,
...
 #if !HAVE_DECL_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
-  shim_set_knownhosts_file(file);
+  shim_set_knownhosts_file (file ?: data->user_known_hosts);
 #endif
   /* HACK: https://bugs.libssh.org/T108 */
   if (!file)
-      g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == SSH_OK);
-  g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_KNOWNHOSTS, file) == SSH_OK);
+    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == SSH_OK);
+  g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_GLOBAL_KNOWNHOSTS, file) == SSH_OK);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: The last commit changes the SSH_OK's to 0's

File: src/ssh/cockpitsshoptions.c [hunk 3]
```diff
@@ -100,8 +99,7 @@ cockpit_ssh_options_from_env (gchar **env)
 {
 
   CockpitSshOptions *options = g_new0 (CockpitSshOptions, 1);
-  options->knownhosts_file = get_environment_val (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE",
-                                                  default_knownhosts);
+  options->knownhosts_file = get_environment_val (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE", NULL);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshoptions.c]: You're right, this patch only works on 0.8.5 and higher.

File: src/ssh/cockpitsshrelay.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/ssh/cockpitsshrelay.c]: Ah, so GitHub just shows them in the wrong order. Nevermind then.

**Comment by martinpitt** [comment]
The ssh options parsing PR landed, so this should now be ready to be rebased and unblocked.

File: src/ssh/cockpitsshoptions.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/ssh/cockpitsshoptions.c]: Note that ubuntu-{1804,stable} have 0.8.{0,1}, with some backports, so we need to get along with these versions for a while.

[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshoptions.c]: Right, but I made the check to default_knownhosts explicit in `set_knownhosts_file`, so it's contained within one version check. So I think that I didn't change behaviour there pre-085

**Comment by Gundersanne** [comment]
forgot the closes, pushed again

**Comment by martinpitt** [comment]
@Gundersanne : For cases like that you can use bots/push-rewrite to get the previous results copied (if there is a net-zero code change)

**Pull request merged**